src/client/dfuse/dfuse_core.c

```diff
@@ -11,38 +11,60 @@
 
 /* Async progress thread.
  *
- * This thread is started at launch time with an event queue and blocks
- * on a semaphore until a asynchronous event is created, at which point
- * the thread wakes up and busy polls in daos_eq_poll() until it's complete.
+ * A number of threads are created at launch, each thread having it's own event queue with a
+ * semaphore to wakeup, posted for each entry added to the event queue and once for shutdown.
+ * When there are no entries on the eq then the thread will yield in the semaphore, when there
+ * are pending events it'll spin in eq_poll() for completion.  All pending events should be
+ * completed before thread exit, should exit be called with pending events.
  */
 static void *
 dfuse_progress_thread(void *arg)
 {
-	struct dfuse_projection_info *fs_handle = arg;
-	int rc;
-	daos_event_t *dev;
-	struct dfuse_event *ev;
+	struct dfuse_eq *eqt = arg;
+	daos_event_t    *dev[128];
+	int              to_consume = 1;
 
 	while (1) {
-		errno = 0;
-		rc = sem_wait(&fs_handle->dpi_sem);
-		if (rc != 0) {
-			rc = errno;
+		int rc;
+		int i;
+
+		for (i = 0; i < to_consume; i++) {
+cont:
+			errno = 0;
+			rc    = sem_wait(&eqt->de_sem);
+
+			if (rc != 0) {
+				rc = errno;
+
+				if (rc == EINTR)
+					D_GOTO(cont, 0);
 
-			if (rc == EINTR)
-				continue;
+				DFUSE_TRA_ERROR(eqt, "Error from sem_wait: %d", rc);
+			}
+		}
+
+		if (eqt->de_handle->dpi_shutdown) {
+			int pending;
+
+			pending = daos_eq_query(eqt->de_eq, DAOS_EQR_ALL, 0, NULL);
+			DFUSE_TRA_INFO(eqt, "There are %d events pending", pending);
 
-			DFUSE_TRA_ERROR(fs_handle, "Error from sem_wait: %d", rc);
+			if (pending == 0)
+				return NULL;
 		}
 
-		if (fs_handle->dpi_shutdown)
-			return NULL;
+		rc = daos_eq_poll(eqt->de_eq, 1, DAOS_EQ_WAIT, 128, &dev[0]);
+		if (rc >= 1) {
+			for (i = 0; i < rc; i++) {
+				struct dfuse_event *ev;
 
-		rc = daos_eq_poll(fs_handle->dpi_eq, 1, DAOS_EQ_WAIT, 1, &dev);
-		if (rc == 1) {
-			daos_event_fini(dev);
-			ev = container_of(dev, struct dfuse_event, de_ev);
-			ev->de_complete_cb(ev);
+				daos_event_fini(dev[i]);
```

Comment:
I wondered about this, yes we do use ev.ev_error after calling daos_event_fini() but the header file is ambiguous about if this is safe and from the code it looks like it is.  ev is actually released (back onto a allocatined but unused descriptor list) in ev->ev_complete_cb() so this would mean duplicating the code into read and write if we moved it however.

What I should be doing is checking the return code here though, but I don't know quite how to handle errors at this point.